src/main/java/com/mraof/minestuck/data/BetterCombatProvider.java

```diff
@@ -0,0 +1,183 @@
+package com.mraof.minestuck.data;
+
+import com.google.gson.*;
+import com.mraof.minestuck.Minestuck;
+import com.mraof.minestuck.item.MSItems;
+import com.mraof.minestuck.item.weapon.MSToolType;
+import com.mraof.minestuck.item.weapon.WeaponItem;
+import net.minecraft.MethodsReturnNonnullByDefault;
+import net.minecraft.data.CachedOutput;
+import net.minecraft.data.DataProvider;
+import net.minecraft.data.PackOutput;
+import net.minecraft.world.item.Item;
+import net.neoforged.neoforge.registries.DeferredItem;
+
+import javax.annotation.ParametersAreNonnullByDefault;
+import java.nio.file.Path;
+import java.util.*;
+import java.util.concurrent.CompletableFuture;
+
+import static com.mraof.minestuck.item.MSItemTypes.*;
+
+@ParametersAreNonnullByDefault
+@MethodsReturnNonnullByDefault
+public class BetterCombatProvider implements DataProvider
+{
+	private final PackOutput output;
+	
+	Map<String, String> weaponWithParent = new HashMap<>();
+	
+	public BetterCombatProvider(PackOutput output)
+	{
+		this.output = output;
+	}
+	
+	/**
+	 * First gets all the weapons from MSItems and adds it to a Map where the key is the registry name and the value is the data.
+	 * Then it checks for custom defined weapons to use either in addition to the existing entries or to replace
+	 */
+	@Override
+	public final CompletableFuture<?> run(CachedOutput cache)
+	{
+		List<CompletableFuture<?>> futures = new ArrayList<>();
+		Path basePath = this.output.getOutputFolder(PackOutput.Target.DATA_PACK).resolve(Minestuck.MOD_ID).resolve("weapon_attributes");
+		
+		MSItems.REGISTER.getEntries().forEach(weaponHolder ->
+		{
+			if(weaponHolder.asOptional().isEmpty())
+				return;
+			
```

Comment:
All entries should already be set up at this point, so I believe this to be redundant.